### ISSUE TITLE
Proposal: List<T>.AsSpan()

### ISSUE THREAD

**Issue by benaadams** [dotnet/corefx#19814]
## Proposed API Additions
```csharp
public class List<T>
{
    public Span<T> AsSpan<T>();
    public ReadOnlySpan<T> AsReadOnlySpan<T>();
}
```
## Sample Usage and Rationale

```csharp
var list = new List<byte>();
foreach(var b in SomeThing)
{
    list.Add(b);
}

var span = list.AsSpan();
`

Arrays are fixed sizes; so creating an array over an unknown sized set involves manual resizing and copying (or upfront over allocation). List already covers this scenario, resizing array behind the scenes.

Proposal is to add a method that returns a correctly sized Span window over the List's array.

**Comment by stephentoub** [comment]
What's the expected behavior in a case like:
```
var list = new List<byte>() { 1, 2, 3, 4 };
var span = list.AsSpan();
list.Add(5); // will cause list's array to be replaced by a bigger one
```
where the span would then be wrapping the "wrong" array?

**Comment by benaadams** [comment]
> What's the expected behavior in a case like:

A warning in docs that `Add`ing or `TrimExcess` to the List may disconnect the Span from live List?

**Comment by mgravell** [comment]
Nice. Would need clear warnings about competing access (as above, but also reminding people that the length won't change if they Remove etc), but: nice idea

**Comment by omariom** [comment]
"snapshot" in the name would be selfdescriptive

**Comment by mgravell** [comment]
Just a minor observation: the `string` equivalent of this ignores `As[Read-only]Span`, and jumps straight to `Slice(...)`, IIRC (I could be misremembering, I'm not in that code).

**Comment by mgravell** [comment]
Related idea: Slice over MemoryStream, encapsulating all the GetBuffer / TryGetBuffer (depending on framework version)

**Comment by omariom** [comment]
Btw, I have a story about `GetBuffer` and how people confuse its meaning.
There is [a mistake in serializing DataTable](https://github.com/dotnet/corefx/blob/master/src/System.Data.Common/src/System/Data/DataSet.cs#L308). I filed a bug 12 or more years ago but MS refused to fix it. I can't remember the exact explanation, something like "BinaryFormatter is not the coolest kid in the block any more" :smile:

**Comment by Tornhoof** [comment]
@stephentoub your last comment in 
```csharp 
foreach (byte b in list) Console.WriteLine(b); // outputs 1, 2, 3, 4, rather than 0, 2, 3, 4, 5
```
is wrong, isn't it?
It should output 1,2,3,4,5 as the list contains 5 now.

For an initial 5 element list it would work as the list currently would have an array size of 8 initially
```csharp
var list = new List<byte>() { 1, 2, 3, 4, 5 };
var span = list.AsSpan();
list.Add(6); // no replacing the array
span[0] = 0; // updates old array
foreach (byte b in list) Console.WriteLine(b); // 0, 2, 3, 4, 5, 6
```
I'd say the API would be too dangerous too put on list, simply because it requires the understanding of the implementation details of list to use fully/correctly.
Maybe only put ``ReadOnlySpan`` on ``ReadOnlyList`` ?

**Comment by omariom** [comment]
> A warning in docs

People don't read the docs since they got intellisense :)

**Comment by mgravell** [comment]
@Tornhoof  by that logic, `MemoryStream` shouldn't have `GetBuffer()` / `TryGetBuffer()`; that has *exactly* the same problems, yet it is a very valuable tool.

I can see immediate use cases for this including serialization and vectorization.

**Comment by mgravell** [comment]
@omariom re that `DataSet` bug: oof, that's a fun catch. I see people doing this *all the time*, sadly (I get a lot of questions about that re protobuf-net; it is  right up there next to people using `MemoryStream` and forgetting to rewind it between usages)

**Comment by Tornhoof** [comment]
@mgravell Yes I admit, it's similar to ``MemoryStream`` and yes ``ReadOnlyList`` does not exist and only ``IReadOnlyList`` exists (I mixed that up somewhere). Your point about contiguous memory is important and only works by relying on the actual implementation details, without some funny interfaces like ``IContiguousMemory`` (and that one implementing the Span API).

**Comment by dadhi** [comment]
May be then `AsDetachedSpan()`, `AsDetachedReadonlySpan()`.

Even if the name itself doesn't click, then the name length should be a warning :-)

**Comment by tom3m** [comment]
AsDetachedSnapshot() AsDetachedReadonlySnapshot()

**Comment by ReubenBond** [comment]
`GetCurrentSpan()` would allude to the potentially transient nature of the result.

**Comment by stephentoub** [comment]
> is wrong, isn't it? It should output 1,2,3,4,5 as the list contains 5 now.

I was highlighting how different folks could have different expectations.  It would be reasonable for someone to expect that since 5 was added to the list and the 0th element was changed to 0 via the span that it would then contain 0,2,3,4,5.  As you point out, the list itself would actually contain 1,2,3,4,5 even though the span would show 0,2,3,4.

> A warning in docs that Adding or TrimExcess to the List may disconnect the Span from live List?

It's not just Adding and TrimExcess that would cause problems.  Consider:
```C#
var list = new List() { 0, 1, 2, 3 };
var span = list.AsSpan();
list.RemoveAt(3);
list.RemoveAt(2);
```
Now the list has a Count of 2 and contains 0,1, but the span still has a length of 4 and contains 0,1,0,0 (though the latter two would depend on whether the list clears elements that don't contain any refs).

I'm simply pointing out it's not a snapshot, nor is it a live representation.  It's some frankenbeast.

**Comment by benaadams** [comment]
To get a real Span snapshot or detached you can use the existing api and do `new Span(list.ToArray())` but that mostly defeats the purpose; both allocating and not being a live copy (for interop), also more apis current exist that take array so you probably wouldn't bother with the Span. 

Though you may do `new ReadOnlySpan(list.ToArray())` to pass a read only array snapshot.

Its a less dangerous api than `MemoryStream` `GetBuffer()` / `TryGetBuffer()` as anything passed the Span can't take a copy as its a stack only type, which they can with the array returned via `GetBuffer()`; which is equally a frankenbeast 😉 

Unlike an array the Span would be more localized in used, due to its stack only nature, so overlapped use between List and Span should be more obvious and `.AsSpan` should be a fairly lightweight non-allocating, operation so isn't a reason to do it at the point of use.

**Comment by mgravell** [comment]
In the spirit of `DangerousGetPinnableReference()`: `DangerousGetSpan()` :) although `GetFrankenspan()` also has a curious appeal...

**Comment by svick** [comment]
@mgravell 

> Additionally, `ReadOnlyCollection<T>` *wraps* a more general `IList<T>` implementation - impossible to get a span for that, as you don't know that the *inner* implementation is contiguous, and even if it is: you don't have access to that data, because it is abstracted.

Is that an argument to add a something like `TryGetSpan` as a default interface method (https://github.com/dotnet/csharplang/issues/52) to `IList<T>` when that becomes possible? It could look something like:

```c#
interface IList<T>
{
    bool TryGetSpan(out Span<T> span)
    {
        span = default;
        return false;
    }

    bool TryGetReadOnlySpan(out ReadOnlySpan<T> span)
    {
        span = default;
        return false;
    }
}

class List<T> : IList<T>
{
    bool TryGetSpan(out Span<T> span)
    {
        span = _items.AsSpan();
        return true;
    }

    bool TryGetReadOnlySpan(out ReadOnlySpan<T> span)
    {
        span = _items.AsReadOnlySpan();
        return true;
    }
}

class ReadOnlyCollection<T> : IList<T>
{
    bool TryGetReadOnlySpan(out Span<T> span)
    {
        return list.TryGetReadOnlySpan(out span);
    }
}
```

**Comment by LYP951018** [comment]
What about:

```cs
void WithSpan(Action<Span<T>> callback)
```
?

**Comment by svick** [comment]
@LYP951018 That doesn't really solve the issue, you can still do e.g.:

```c#
list.WithSpan(span =>
{
    list.Add(42); // or whatever

    // span is now "frankenspan"
});
```

**Comment by LYP951018** [comment]
OMG...

**Comment by mgravell** [comment]
Also, most of the time when the span would be useful, it would be extremely inconvenient to have to force the work into a delegate, plus that has overhead of the delegate instance, the capture context, and the virtual dispatch.

If we're going to end up with frankenspan, we might as well do it efficiently :)

**Comment by benaadams** [comment]
Would ReadOnlyFrankenSpan be better?

**Comment by wanton7** [comment]
How about adding explicit interface like this? This could be added to StringBuilder as well.

```
interface IDangerousArrayAccess<T>
{
    T[] Array { get; }
    int Count { get; }
}
```

**Comment by mgravell** [comment]
> How about adding explicit interface like this?

Adding interfaces is much harder to deploy, unless we get "interface implementation via extensions"; can't see it happening, and even if it did: would be restricted to certain targets.

> This could be added to StringBuilder as well.

StringBuilder doesn't use a single array origin; it uses a chunk chain IIRC - and even that is an implementation detail. IIRC it used to use an oversized `string` that it mutated in-place!

**Comment by Bobris** [comment]
Did somebody solve this by adding `Add` extension method on top of `Memory<T>`? Or just implemented own `struct List`?

**Comment by yahorsi** [comment]
.20 cents.
Currently List makes an "additional bounds check" to the one array does. In other words when accessing List indexer or list iterator it must make sure the index given is less than list size. That is relatively big overhead. If we could have AsSpan or even AsReadOnlySpan this will help to significantly improve list access perf. And, even, potentically could result in complete array bounds check elimination vs current state when we have actually 2 bounds checks.
